learntools/machine_learning/ex3.py

```diff
@@ -1,42 +1,77 @@
 from learntools.core.richtext import CodeSolution as CS
 from learntools.core.problem import *
 
 
-class SetTarget(EqualityCheckProblem):
+class SetTarget(CodingProblem):
     _var = 'y'
-    _hint = "Use the `pd.read_csv` function"
-    _solution = CS('home_data = pd.read_csv(iowa_file_path)')
+    _hint = ("Use `print(home_data.columns)`. The column you want is at the end "
+            "of the list. Use the dot notation to pull out this column from the DataFrame")
+    _solution = CS('y = home_data.SalePrice')
+
+    def check(self, targ):
+        assert isinstance(targ, pd.Series), ("`home_data` should be a Pandas Series "
+                                             "with the actual data. Your current "
+                                             "answer is a `{}`").format(type(targ),)
+        true_mean = 180921.19589041095
+        assert int(targ.mean()) == int(true_mean), ("You've selected the wrong data.")
+
 
+class SelectPredictionData(CodingProblem):
+    _var = 'X'
+    _hint = ("Capitalization and spelling are important when specifying variable names. "
+             "Use the brackets notation when data for X.")
+    _solution = CS(
+"""predictor_names = ["LotArea", "YearBuilt", "1stFlrSF", "2ndFlrSF",
+                      "FullBath", "BedroomAbvGr", "TotRmsAbvGrd"]
+
+X=home_data[predictor_names]""")
     def check(self, df):
-        assert isinstance(df, pd.DataFrame), ("`home_data` should be a DataFrame,"
+        assert isinstance(df, pd.DataFrame), ("`X` should be a DataFrame,"
                 " not `{}`").format(type(df),)
-        expected_shape = (1460, 81)
+        expected_shape = (1460, 7)
         assert df.shape == expected_shape, ("Expected {} rows and {} columns, but"
                 " got shape {}").format(expected_shape[0], expected_shape[1], df.shape)
 
-class HomeDescription(EqualityCheckProblem):
-    _vars = ['ave_lot_size', 'newest_home_age']
-    _expected = [10517, 8]
-    _hint = 'Run the describe command. Lot size is in the column called LotArea. Also look at YearBuilt'
-    _solution = CS(
-"""# using data read from home_data.describe()
-avg_lot_size = 10517
-newest_home_age = 8
 
-    """)
+class CreateModel(CodingProblem):
+    _var = 'iowa_model'
+    _hint = ("Include random_state when specifying model. Data is specified when fitting it.")
+    _solution = CS("""from sklearn.tree import DecisionTreeRegressor
+iowa_model = DecisionTreeRegressor(random_state=1)
+iowa_model.fit(X, y)""")
+
+    def check(self, dtree):
+        # Not checking what they fit, because likely mistakes cause exceptions
+
+        assert type(dtree) == sklearn.tree.tree.DecisionTreeRegressor, \
+                    ("Expected type object of type DecisionTreeRegressor but got an "
+                     "object of type {}").format(type(dtree))
+        assert dtree.random_state is not None, "You forgot to set the random_state."
+        assert dtree.tree_, "You have not fit the model."
```

Comment:
What version of sklearn do you have running locally.  I'm inclined not to worry about this if yours is old. tree_ exists after the model has been fit in 0.18.1

Reply:
If this is a big improvement, I don't see it yet:

If they have a newish version of sklearn, I don't think it has any effect.

If they have an older version of sklearn, we'll always claim they failed to fit the model, even if their code does fit the model.  That might be even more frustrating than an incomprehensible stack trace.

fwiw, I'm on sklearn `0.18.1` and Kaggle is on `0.20.dev0`. Both those have `tree_` on fitted models.  This makes me think it's only older versions that don't have this.  

I'm skeptical learntools will be run outside kaggle, and Kaggle is unlikely to go back to an older version of sklearn.  So I'm inclined to think we'll get by without much practical problem no matter where we land on this.

Reply:
> fwiw, I'm on sklearn 0.18.1 and Kaggle is on 0.20.dev0. Both those have tree_ on fitted models.

I'm on 0.19.1, which also has `tree_` on fitted models. The (minor) issue I was referring to here was for *unfitted* models. This code assumes that an unfitted model will have a `tree_` attr which is falsey. But when I tried checking an unfitted model, I got an AttributeError because my model had no `tree_` attr.

Reply:
Gotcha.  Will make the fix you suggested.